### PR TITLE
(#528) Allow running tasks on non-AIO nodes

### PR DESCRIPTION
**Pull request by smortex**

### CODE REVIEW

File: lib/mcollective/util/tasks_support.rb [hunk 1]
```diff
@@ -59,34 +59,54 @@ def puppet_type_to_ruby(type)
...
-          "/opt/puppetlabs/puppet/bin"
+      # Path to look for binaries like wrappers etc
+      #
+      # @return [Array<String>]
+      def search_path
+        @search_path ||= [
+          'C:\Program Files\Puppet Labs\Puppet\bin',
+          "/opt/puppetlabs/puppet/bin",
+          "/usr/local/bin"
+        ].select { |path| File.exist?(path) && File.directory?(path) }
```
[the remark this answers is not in the document]
Comment (smortex) [reply lib/mcollective/util/tasks_support.rb]: Wow, I believed you needed to be an admin to add files outside your "personal" directory in recent versions of Windows (I mean something better than Windows 98)!

Making this configurable is probably the best way to handle this then.  I guess that converting this file to a template is a terrible idea (`task_support.rb.erb`), so maybe we can use a template to drop some kind of configuration file next to the .rb file (`task_support.{conf,yaml,ini,json}`), and read this file from `task_support.rb`?  The data could come from Hiera, and overwriting should be possible for some advanced use-cases (e.g. Debian ARM where AIO is not available… I have not looked into this yet but guess there are dragons there).

Does this look reasonable?  If so do you have a preference regarding the configuration file format?

I should be able to work on this after the week-end :-)

Thanks for the feedback!

[the remark this answers is not in the document]
Comment (ripienaar) [reply lib/mcollective/util/tasks_support.rb]: Perhaps, but who knows - i dont have a windows I am not an admin on handy, best be safe.

So to make it configurable do something like this:

```
def wrapper_path
  choria.get_option("choria.tasks.wrapper_path", "/opt/puppetlabs/puppet/bin/task_wrapper")
end
```

then we can just add hiera data for the various operating systems:

```
mcollective_choria::config:
  tasks.wrapper_path: /whatever
```

**Comment by smortex** [comment]
Done!  I renamed a few functions to make it clear that they are tight to the AIO paths (`#bin_path`, `#wrapper_path`)

File: spec/unit/mcollective/util/tasks_support_spec.rb [hunk 2]
```diff
@@ -482,19 +482,19 @@ module Util
...
+          ts.stubs(:aio_bin_path).returns("/nonexisting/bin")
           expect(ts.ps_shim_path).to eq("/nonexisting/bin/PowershellShim.ps1")
         end
       end
 
       describe "#bin_path" do
         it "should support windows" do
           Util.stubs(:windows?).returns(true)
-          expect(ts.bin_path).to eq('C:\Program Files\Puppet Labs\Puppet\bin')
+          expect(ts.aio_bin_path).to eq('C:\Program Files\Puppet Labs\Puppet\bin')
```
Comment (ripienaar) [inline spec/unit/mcollective/util/tasks_support_spec.rb]: gonna merge this but I've since learned this is the wrong path, will fix in a follow up PR

**Comment by ripienaar** [comment]
Thank you @smortex

File: spec/unit/mcollective/util/tasks_support_spec.rb (hunk 2 above)
Comment (smortex) [reply spec/unit/mcollective/util/tasks_support_spec.rb]: I guess you are referring to #530.  I applied the suggested change to the Windows path and tested on a Windows machine, but did I was still not able to run a task (I mean that changing the path seems to not be enough).  Hacking on Windows is probably the worst experience I can afford, and my test environment is not handy at all, so I could not dig deeper into fixing this issue.

However, if someone is to propose a fix, I should be able to try it in a timely fashion.  Feel free to ping me if I can help!

**Pull request merged**